codegen/debug.go

```diff
@@ -504,3 +539,155 @@ func attr(dgst digest.Digest, op pb.Op) (string, string) {
 		return dgst.String(), "plaintext"
 	}
 }
+
+func Exec(ctx context.Context, cln *client.Client, fs Filesystem, r io.ReadCloser, w io.Writer, args ...string) error {
+	if len(args) == 0 {
+		args = []string{"/bin/sh"}
+	}
+
+	s, err := llbutil.NewSession(ctx, fs.SessionOpts...)
+	if err != nil {
+		return err
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		return s.Run(ctx, cln.Dialer())
+	})
+
+	g.Go(func() error {
+		var pw progress.Writer
+
+		mw := MultiWriter(ctx)
+		if mw != nil {
+			pw = mw.WithPrefix("", false)
+		}
+
+		return solver.Build(ctx, cln, s, pw, func(ctx context.Context, c gateway.Client) (res *gateway.Result, err error) {
+			def, err := fs.State.Marshal(ctx, llb.LinuxAmd64)
+			if err != nil {
+				return
+			}
+
+			res, err = c.Solve(ctx, gateway.SolveRequest{
+				Definition: def.ToPB(),
+			})
+			if err != nil {
+				return
+			}
+
+			ctr, err := c.NewContainer(ctx, gateway.NewContainerRequest{
+				Mounts: []gateway.Mount{{
```

Comment:
Out of scope for this PR, but wonder if we should respect mounts in the with-options if the cursor is on a `run` call.

Also should we respect existing state like env vars set on `llb.State`, current working directory, etc?

Reply:
I've added support for this in the latest commit. In addition to respecting those options when the cursor is on a `run` statement, I've added equivalent options for `breakpoint` (for example, to exec a shell with specific mounts available).